wordpress-mobile/WordPress-iOS
# Fix media uploads error handling

Part of https://github.com/wordpress-mobile/WordPress-iOS/issues/21190. 

Fixes missing video upload limits reported by @momo-ozawa [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/21398#pullrequestreview-1593143993). The fix requires a lot of changes because I found multiple other production issues on the Media screen that had to be fixed.

@osullivanchris, if you have a better idea of how to handle this scenario, given the technical limitations, please let me know. It may also be worth re-visiting this entire flow in the scope of the Media changes.

## Handling Upload Limits

One of the advantages of using a custom picker is that it allows us to support scenarios like the upload limits. The custom picker displays the errors in the picker itself ([recording.mp4](https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/55cbcd84-3707-49a9-b57c-9732e7995fb5)).

<img width="320" alt="current-screenshot" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/549bbfe2-6780-4c7b-9fcb-338b9e8bfee4">

Unfortunately, it wasn't possible with `PHPickerViewController`. I implemented it by handling this as other export errors ([recording.mp4](https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/e543616d-4775-47c1-b636-e31d59a8ad3c))

<img width="320" alt="Screenshot 2023-08-24 at 4 37 39 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/03d5720d-5ef5-4416-9a1a-86cd09424148">  <img width="320" alt="Screenshot 2023-08-24 at 4 53 39 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/f050e93f-524a-48e9-b731-2d15f6f74c37">


To implement it, I added duration limit checks to `MediaVideoExporter`. It was supposed to just work, but I encountered a few issues I had to fix.

### 1. Fix slow exports from NSItemProvider (https://github.com/wordpress-mobile/WordPress-iOS/commit/0376ca171faae2b0aa46a814b9938e2f894b02c7)

The `NSItemProvider` returned by `PHPickerViewController` doesn't return the video duration – you have to load the file first. But `loadFileRepresentation` was too slow: it was taking more than a minute for my 500 MB video. The reason was transcoding.

I followed the [advice](https://developer.apple.com/forums/thread/652496) of an Apple engineer to disable transcoding on `PHPickerViewController`. The `loadFileRepresentation` was now taking around 200 ms. It’s the correct course of action because [MediaVideoExporter.swift](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/Utility/Media/MediaVideoExporter.swift) also performs its own transcoding. We definitely don’t want to do it twice. 

### 2. Fix an issue with MediaImportService not reporting errors (https://github.com/wordpress-mobile/WordPress-iOS/commit/13285a395743dd3aaec2bc9a12bf03199e9a5286)

`MediaImportService` was incorrectly calling a completion closure with `.success` on export failures. The errors were never reported to the Media screen.

### 3. Fix invalid localizedDescription of exporter errors (https://github.com/wordpress-mobile/WordPress-iOS/commit/c832575dbfed8b8ce4d1a5bf1f9aa8f141795020)

Before -> After

<img width="320" alt="Screenshot 2023-08-24 at 3 23 45 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/be4f77db-9496-4bd0-a719-b3f6140a039b"> <img width="320" alt="Screenshot 2023-08-24 at 4 37 39 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/a9b57f3b-47ec-4997-a107-2f05925e52ef">

### 4. Fix incorrect error banner displayed when upload fails https://github.com/wordpress-mobile/WordPress-iOS/commit/c144f66e4bd6cb250c0af1c523fb19db44c57d0e

Before -> After

<img width="320" alt="Screenshot 2023-08-24 at 3 42 15 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/d25253c5-54da-45a1-b12f-7c237b732e58"> <img width="320" alt="Screenshot 2023-08-24 at 3 47 59 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/29e8eb4c-84d9-45e8-bd75-32214f38f1e3">

## To test:

- Add a long video (5 minutes to Photos)
- Select a site with a free plan
- Open Media and try uploading the video
- Verify that the error is shown

Ideally we also need to re-run the tests from https://github.com/wordpress-mobile/WordPress-iOS/pull/21343

## Regression Notes

1. Potential unintended areas of impact: Media
2. What I did to test those areas of impact (or what existing automated tests I relied on): manual
3. What automated tests I added (or what prevented me from doing so): n/a

PR submission checklist:


UI Changes testing checklist:
